entity-service-impl/src/main/java/org/hypertrace/entity/service/util/DocStoreConverter.java

```diff
@@ -84,28 +89,108 @@ public static Filter getTenantIdEqFilter(String tenantId) {
   }
 
   private static Filter transform(AttributeFilter filter) {
+    try {
+      if (filter.hasAttributeValue()) {
+        if (ATTRIBUTES_LABELS_FIELD_NAME.equals(filter.getName()) && filter.getOperator() == Operator.EQ) {
+          return transformToEqFilterWithValueListRhs(filter);
+        } else if (ATTRIBUTES_LABELS_FIELD_NAME.equals(filter.getName()) && filter.getOperator() == Operator.IN) {
+          return transformToOrFilterChainForStrArray(filter);
+        } else {
+          return transformNonListRhsFilterTypes(filter);
+        }
+      } else { // AND or OR filter chains
+        Filter f = new Filter();
+        f.setFieldName(filter.getName());
+        f.setOp(transform(filter.getOperator()));
+
+        f.setChildFilters(
+            filter.getChildFilterList().stream()
+                .map(DocStoreConverter::transform)
+                .collect(Collectors.toList())
+                .toArray(new Filter[]{}));
+        return f;
+      }
+    } catch (IOException ioe) {
+      throw new IllegalArgumentException(String.format("transform: Error converting filter for query. Filter:  %s", filter), ioe);
+    }
+  }
+
+  private static Filter transformNonListRhsFilterTypes(AttributeFilter filter) {
     try {
       Filter f = new Filter();
       f.setFieldName(filter.getName());
       f.setOp(transform(filter.getOperator()));
-      if (filter.hasAttributeValue()) {
-        transform(filter.getAttributeValue(), f, isPartOfAttributeMap(filter.getName()));
-      }
-      f.setChildFilters(
-          filter.getChildFilterList().stream()
-              .map(DocStoreConverter::transform)
-              .collect(Collectors.toList())
-              .toArray(new Filter[]{}));
+      transform(filter.getAttributeValue(), f, isPartOfAttributeMap(f.getFieldName()));
+
+      // Set child filters to empty string
+      f.setChildFilters(new Filter[]{});
       return f;
     } catch (IOException ioe) {
-      throw new IllegalArgumentException("Error converting filter for query");
+      throw new IllegalArgumentException(String.format("transformNonListRhsFilterTypes: Error converting filter for query. Filter:  %s", filter), ioe);
     }
   }
 
   private static boolean isPartOfAttributeMap(String fieldName) {
     return !EntityServiceConstants.ENTITY_CREATED_TIME.equalsIgnoreCase(fieldName);
   }
 
+  private static Filter transformToOrFilterChainForStrArray(AttributeFilter attributeFilter) {
+    String fieldName = attributeFilter.getName() + VALUE_LIST_VALUES_CONST;
+
+    Filter f = new Filter();
+    f.setFieldName("");
+    f.setOp(Op.OR);
+
+    f.setChildFilters(
+        attributeFilter.getAttributeValue().getValueList().getValuesList().stream()
+            .map(rhsAttributeValue -> {
+              Filter childFilter = new Filter();
```

Comment:
how's this block different from the below `transformToEqFilterWithValueListRhs()`? In that method also you seem to be using only AttributeValue. Can you reuse the code probably?

Reply:
Mmh let me take a look. There is a reason in the back of my head but may be not.

Reply:
No reason actually. Will refactor.